proposals/gc/MVP.md

```diff
@@ -97,20 +118,167 @@ New abbreviations are introduced for reference types in binary and text format,
   - `storagetype ::= <valtype> | <packedtype>`
   - `packedtype ::= i8 | i16`
 
+TODO: Need to be able to use `i31` as a type definition.
+
+
+#### Type Contexts
+
+Validity of a module is checked under a context storing the definitions for each type. In the case of recursive types, this definition is given by a respective projection from the full type:
+```
+ctxtype ::= <deftype> | (rec <deftype>*).<i>
+```
+
+#### Auxiliary Definitions
+
 * Unpacking a storage type yields `i32` for packed types, otherwise the type itself
   - `unpacked(t) = t`
   - `unpacked(pt) = i32`
 
-TODO: Need to be able to use `i31` as a type definition.
+* Context lookup `C($t)` is extended to recursive type indices by ignoring the `rec` marker.
+  - `C(rec $t) = C($t)`
+
+* Unrolling a possibly recursive context type projects the respective item
+  - `unroll($t)                 = unroll(<ctxtype)`  iff `C($t) = <ctxtype>`
+  - `unroll(<deftype>)          = <deftype>`
+  - `unroll((rec <deftype>*).i) = (<deftype>*)[i]`
+
+* Expanding a type definition unrolls it and returns its plain definition
+  - `expand($t)                 = expand(<ctxtype)`  iff `C($t) = <ctxtype>`
+  - `expand(<ctxtype>) = <strtype>`
+    - where `unroll(<ctxttype>) = sub x* <strtype>`
+
+
+
+#### Type Validity
+
+Some of the rules define a type as `ok` for a certain index, written `ok(x)`. This controls uses of type indices as supertypes inside a recursive group: the subtype hierarchy must not be cyclic, and hence any type index used for a supertype is required to be smaller than the index `x` of the current type.
+
+* a sequence of type definitions is valid if each item is valid within the context containing the prior items
+  - `<typedef0> <typedef>* ok`
+    - iff `<typedef0> ok` and extends the context accordingly
+    - and `<typedef>* ok` under the extended context
+
+* a plain type definition is valid if its `deftype` is at its type index
+  - `<deftype> ok` and extends the context with `<deftype>`
+    - iff `<deftype> ok($t)` where `$t` is the next unused (i.e., current) type index
+
+* a recursive type definition is valid if its types are valid under the context containing all of them
+  - `rec <deftype>* ok` and extends the context with `<ctxtype>*`
+    - iff `<deftype>* ok($t)` under the extended context(!)
+    - where `x` is the next unused (i.e., current) type index
+    - and `N = |<deftype>*|-1`
+    - and `<deftype'>* = <deftype>*[$t:=rec $t, ..., $t+N:=rec $t+N]`
```

Comment:
Should this read:
    - and `<deftype'>* = <deftype>*[$x:=rec $x, ..., $x+N:=rec $x+N]`
?